pkg/gce-cloud-provider/compute/gce-compute.go

```diff
@@ -456,3 +474,116 @@ func (cloud *CloudProvider) GetInstanceOrError(ctx context.Context, instanceZone
+	op, err := cloud.service.Snapshots.Delete(cloud.project, snapshotName).Context(ctx).Do()
+	if err != nil {
+		if IsGCEError(err, "notFound") {
+			// Already deleted
+			return nil
+		}
+		return err
+	}
+	err = cloud.waitForGlobalOp(ctx, op)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (cloud *CloudProvider) CreateSnapshot(ctx context.Context, volKey *meta.Key, snapshotName string) (*compute.Snapshot, error) {
+	switch volKey.Type() {
+	case meta.Zonal:
+		snapshot, err := cloud.createSnapshotOfZonalDisk(ctx, volKey, snapshotName)
+		return snapshot, err
+	case meta.Regional:
+		snapshot, err := cloud.createSnapshotOfRegionalDisk(ctx, volKey, snapshotName)
+		return snapshot, err
+	default:
+		return nil, fmt.Errorf("could not create snapshot, key was neither zonal nor regional, instead got: %v", volKey.String())
+	}
+}
+
+func (cloud *CloudProvider) createSnapshotOfZonalDisk(ctx context.Context, volKey *meta.Key, snapshotName string) (*compute.Snapshot, error) {
+	snapshotToCreate := &compute.Snapshot{
+		Name:        snapshotName,
+	}
+
+	_, err := cloud.service.Disks.CreateSnapshot(cloud.project, volKey.Zone, volKey.Name, snapshotToCreate).Context(ctx).Do()
+
+	if err != nil {
+		if IsGCEError(err, "alreadyExists") {
+			snapshot, err := cloud.GetSnapshot(ctx, snapshotName)
+			if err != nil {
+				return nil, err
+			}
+			glog.Warningf("GCE PD snapshot %s already exists, reusing", snapshotName)
+			return snapshot, nil
+		}
+		return nil, fmt.Errorf("unkown disk create snapshot error: %v", err)
+	}
+
+	return cloud.waitForSnapshotCreation(ctx, snapshotName)
+}
+
+func (cloud *CloudProvider) createSnapshotOfRegionalDisk(ctx context.Context, volKey *meta.Key, snapshotName string) (*compute.Snapshot, error) {
+	snapshotToCreate := &computebeta.Snapshot{
+		Name:        snapshotName,
+	}
+
+	_, err := cloud.betaService.RegionDisks.CreateSnapshot(cloud.project, volKey.Region, volKey.Name, snapshotToCreate).Context(ctx).Do()
+	if err != nil {
+		if IsGCEError(err, "alreadyExists") {
+			snapshot, err := cloud.GetSnapshot(ctx, snapshotName)
+			if err != nil {
+				return nil, err
+			}
+			glog.Warningf("GCE PD snapshot %s already exists, reusing", snapshotName)
+			return snapshot, nil
+		}
+		return nil, fmt.Errorf("unkown disk create snapshot error: %v", err)
+	}
+
+	return cloud.waitForSnapshotCreation(ctx, snapshotName)
+
+}
+
+func (cloud *CloudProvider) waitForSnapshotCreation(ctx context.Context, snapshotName string) (*compute.Snapshot, error) {
+	ticker := time.NewTicker(time.Second)
+	defer ticker.Stop()
+	timer := time.NewTimer(waitForSnapshotCreationTimeOut)
+	defer timer.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			glog.V(5).Infof("Checking GCE Snapshot %s.", snapshotName)
+			snapshot, err := cloud.GetSnapshot(ctx, snapshotName)
+			if err != nil {
+				glog.Warningf("Error in getting snapshot %s, %v", snapshotName, err)
+			} else if snapshot != nil {
+				if snapshot.Status != "CREATING" {
```

Comment:
instead of negative check for `CREATING` is there something specific we're looking for like `READY`?

Reply:
The status could be CREATING, DELETING, FAILED, READY, or UPLOADING. Except CREATING, all other states mean that snapshot is already been taken or failed to create.